File: services/ats-system-checker/src/controllers/job.controller.py
"""
Job Controller for ATS System
Handles job posting creation, retrieval, and management
"""

from typing import Optional, List
from datetime import datetime
from fastapi import HTTPException, UploadFile
import httpx
import os
import uuid

from src.utils.logging_config import get_logger
from src.utils.responses import success_response
from src.models.api_models import JobPostingResponse, ApplicationsListResponse, ApplicationListResponse
from src.services.database_service import DatabaseService
from src.services.evaluation_service import EvaluationService
from src.services.s3_service import S3Service
from src.config.settings import get_settings

logger = get_logger(__name__)
settings = get_settings()


class JobController:
    """Controller for job posting operations"""
    
    def __init__(
        self,
        database_service: DatabaseService,
        evaluation_service: EvaluationService,
        s3_service: S3Service,
    ):
        """
        Initialize job controller
        
        Args:
            database_service: Database service instance
            evaluation_service: Evaluation service instance
            s3_service: S3 service instance
        """
        self.database_service = database_service
        self.evaluation_service = evaluation_service
        self.s3_service = s3_service
        
        # Debug: Log S3 service status on initialization
        logger.info(f"JobController initialized with S3 service: {self.s3_service.is_configured()}")
        logger.info(f"S3 bucket: {self.s3_service.bucket_name}")
        logger.info(f"S3 client ready: {self.s3_service.s3_client is not None}")
    
    async def create_job_posting(
        self,
        title: str,
        description: str,
        required_skills: str,
        additional_details: Optional[str] = None,
        hr_email: Optional[str] = None,
        hr_name: Optional[str] = None,
        evaluation_threshold: int = 70,
        quiz_required: bool = True,
        quiz_pass_threshold: int = 7
    ) -> JobPostingResponse:
        """
        Create a new job posting and return shareable link
        
        Args:
            title: Job title
            description: Job description
            required_skills: Required skills (comma-separated)
            additional_details: Additional job details
            hr_email: HR contact email
            hr_name: HR contact name
            evaluation_threshold: Minimum CV score for acceptance
            quiz_required: Whether quiz is required for accepted candidates
            quiz_pass_threshold: Minimum quiz score to pass
            
        Returns:
            JobPostingResponse: Job posting details with shareable link
        """
        try:
            logger.info(f" Creating new job posting: {title}")
            
            # Validation
            if not title.strip() or not description.strip() or not required_skills.strip():
                raise HTTPException(
                    status_code=400,
                    detail="Job title, description, and required skills are required"
                )
            
            # Parse required skills
            skills_list = [skill.strip() for skill in required_skills.split(',') if skill.strip()]
            
            # Create job posting
            job_posting = await self.database_service.create_job_posting(
                title=title,
                description=description,
                required_skills=skills_list,
                additional_details=additional_details,
                hr_email=hr_email,
                hr_name=hr_name,
                evaluation_threshold=evaluation_threshold,
                quiz_required=quiz_required,
                quiz_pass_threshold=quiz_pass_threshold
            )
            
            # Generate shareable link
            base_url = settings.get_frontend_url()
            shareable_link = f"{base_url}/apply/{job_posting.job_id}"
            
            logger.info(f" Created job posting with ID: {job_posting.job_id}")
            
            return JobPostingResponse(
                job_id=job_posting.job_id,
                title=job_posting.title,
                description=job_posting.description,
                required_skills=job_posting.required_skills,
                shareable_link=shareable_link,
                created_at=job_posting.created_at,
                is_active=job_posting.is_active
            )
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f" Error creating job posting: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to create job posting: {str(e)}")
    
    async def get_job_posting(self, job_id: str) -> JobPostingResponse:
        """
        Get job posting details by job ID
        
        Args:
            job_id: Job posting ID
            
        Returns:
            JobPostingResponse: Job posting details
        """
        try:
            logger.info(f" Retrieving job posting: {job_id}")
            
            job_posting = await self.database_service.get_job_posting_by_id(job_id)
            
            if not job_posting:
                raise HTTPException(
                    status_code=404,
                    detail="Job posting not found"
                )
            
            if not job_posting.is_active:
                raise HTTPException(
                    status_code=410,
                    detail="This job posting is no longer active"
                )
            
            # Generate shareable link
            base_url = settings.get_frontend_url()
            shareable_link = f"{base_url}/apply/{job_posting.job_id}"
            
            return JobPostingResponse(
                job_id=job_posting.job_id,
                title=job_posting.title,
                description=job_posting.description,
                required_skills=job_posting.required_skills,
                shareable_link=shareable_link,
                created_at=job_posting.created_at,
                is_active=job_posting.is_active
            )
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f" Error retrieving job posting: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to retrieve job posting: {str(e)}")
    
    async def get_all_job_postings(self, include_inactive: bool = False) -> List[JobPostingResponse]:
        """
        Get all job postings
        
        Args:
            include_inactive: Whether to include inactive job postings
            
        Returns:
            List[JobPostingResponse]: List of job postings
        """
        try:
            logger.info(f" Getting all job postings (include_inactive: {include_inactive})")
            
            job_postings = await self.database_service.get_all_job_postings(include_inactive)
            
            response_data = []
            for job in job_postings:
                job_response = JobPostingResponse(
                    job_id=job.job_id,
                    title=job.title,
                    description=job.description,
                    required_skills=job.required_skills,
                    shareable_link=f"{settings.get_frontend_url()}/apply/{job.job_id}",
                    created_at=job.created_at,
                    is_active=job.is_active
                )
                response_data.append(job_response)
            
            logger.info(f" Successfully retrieved {len(response_data)} job postings")
            return response_data
            
        except Exception as e:
            logger.error(f" Failed to get job postings: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to get job postings: {str(e)}")
    
    async def submit_application(
        self,
        job_id: str,
        cv_file: UploadFile,
        candidate_email: str,
        candidate_name: str
    ):
        """
        Submit application for a job posting
        
        Args:
            job_id: Job posting ID
            cv_file: Uploaded CV file
            candidate_email: Candidate email address (required)
            candidate_name: Candidate full name (required)
            
        Returns:
            APIResponse: Application submission confirmation
        """
        try:
            logger.info(f" Processing application for job: {job_id}")
            
            # Get job posting
            job_posting = await self.database_service.get_job_posting_by_id(job_id)
            if not job_posting:
                raise HTTPException(status_code=404, detail="Job posting not found")
            
            if not job_posting.is_active:
                raise HTTPException(status_code=410, detail="This job posting is no longer active")
            
            # Validate file
            if not cv_file or not cv_file.filename or cv_file.filename == '':
                raise HTTPException(status_code=400, detail="Please upload a valid CV file")
            
            if not cv_file.filename.lower().endswith(('.pdf', '.docx')):
                raise HTTPException(status_code=400, detail="CV must be in PDF or DOCX format")
            
            # SECURITY CHECK: Prevent duplicate applications by email
            existing_application = await self.database_service.check_duplicate_email_application(candidate_email, job_id)
            if existing_application:
                raise HTTPException(
                    status_code=409,
                    detail=f"An application with email '{candidate_email}' has already been submitted for this job posting. Each candidate can only apply once per job."
                )
            
            # Try to upload file to S3, fallback to local storage if S3 fails
            cv_content = None
            use_local_storage = False
            file_url = None
            s3_key = None
            
            # Debug: Check S3 service status
            logger.info(f"S3 service configured: {self.s3_service.is_configured()}")
            logger.info(f"S3 service client ready: {self.s3_service.s3_client is not None}")
            logger.info(f"S3 bucket name: {self.s3_service.bucket_name}")
            logger.info(f"S3 region: {self.s3_service.region}")
            
            try:
                if self.s3_service.is_configured():
                    logger.info("S3 service is configured, attempting S3 upload...")
                    try:
                        logger.info("About to call S3 upload_file method...")
                        s3_result = await self.s3_service.upload_file(cv_file)
                        logger.info(f"S3 upload successful! Result: {s3_result}")
                        file_url = s3_result['file_url']
                        s3_key = s3_result['s3_key']
                        # Read file content for AI processing after S3 upload
                        await cv_file.seek(0)
                        cv_content = await cv_file.read()
                        logger.info(f"File successfully uploaded to S3: {file_url}")
                        logger.info(f"S3 key: {s3_key}")
                        logger.info(f"File URL to be stored: {file_url}")
                        # Skip local storage logic since S3 upload succeeded
                        use_local_storage = False
                        logger.info("S3 upload completed successfully, skipping local storage")
                    except HTTPException as s3_error:
                        logger.warning(f"S3 upload failed with HTTPException: {s3_error.detail}, falling back to local storage")
                        use_local_storage = True
                        await cv_file.seek(0)  # Reset file position for local storage
                    except Exception as s3_error:
                        logger.warning(f"S3 upload failed with Exception: {str(s3_error)}, falling back to local storage")
                        logger.warning(f"Exception type: {type(s3_error)}")
                        logger.warning(f"Exception details: {s3_error}")
                        use_local_storage = True
                        await cv_file.seek(0)  # Reset file position for local storage
                else:
                    logger.warning("S3 not configured, using local storage")
                    use_local_storage = True
                
                # Only execute local storage logic if S3 failed or wasn't configured
                if use_local_storage:
                    # Fallback to local storage
                    logger.info("Using local storage for file upload")
                    
                    # Validate file has a name
                    if not cv_file.filename:
                        raise HTTPException(status_code=400, detail="File must have a filename")
                    
                    # Generate unique filename
                    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                    unique_id = str(uuid.uuid4())[:8]
                    file_extension = os.path.splitext(cv_file.filename)[1].lower()
                    
                    # Ensure we have a valid extension
                    if not file_extension:
                        file_extension = '.pdf'  # Default to PDF if no extension
                    
                    unique_filename = f"cv_{timestamp}_{unique_id}{file_extension}"
                    
                    # Save to local uploads directory
                    upload_path = os.path.join(settings.UPLOAD_FOLDER, unique_filename)
                    
                    # Ensure directory exists
                    try:
                        os.makedirs(settings.UPLOAD_FOLDER, exist_ok=True)
                        logger.info(f"Created/verified upload directory: {settings.UPLOAD_FOLDER}")
                    except Exception as dir_error:
                        logger.error(f"Failed to create upload directory: {dir_error}")
                        raise HTTPException(status_code=500, detail=f"Failed to create upload directory: {str(dir_error)}")
                    
                    # Read file content for both saving and AI processing
                    try:
                        cv_content = await cv_file.read()
                        if not cv_content:
                            raise HTTPException(status_code=400, detail="File is empty")
                        
                        with open(upload_path, 'wb') as f:
                            f.write(cv_content)
                        
                        logger.info(f"File saved locally: {upload_path} ({len(cv_content)} bytes)")
                    except Exception as write_error:
                        logger.error(f"Failed to write file: {write_error}")
                        raise HTTPException(status_code=500, detail=f"Failed to save file: {str(write_error)}")
                    
                    # For local storage, construct the full URL
                    file_url = f"{settings.UPLOADS_BASE_URL}/uploads/{unique_filename}"
                    s3_key = unique_filename
                    
                    logger.info(f"File uploaded successfully to local storage: {unique_filename}")
            except Exception as upload_error:
                logger.error(f"File upload failed: {upload_error}")
                raise HTTPException(
                    status_code=500, 
                    detail=f"File upload failed: {str(upload_error)}"
                )
            
            if not cv_content or len(cv_content) < 100:
                raise HTTPException(status_code=400, detail="Uploaded file is empty or too small.")
            
            # Extract text from CV using AI service
            try:
                async with httpx.AsyncClient() as client:
                    files = {'file': (cv_file.filename, cv_content)}
                    ai_url = f"{settings.AI_SERVICE_URL}/extract-text"
                    response = await client.post(ai_url, files=files, timeout=60)
                    try:
                        response.raise_for_status()
                    except httpx.HTTPStatusError as http_exc:
                        logger.error(f"AI service error: {response.text}")
                        raise HTTPException(status_code=500, detail=f"AI service error: {response.text}")
                    data = response.json()
                    cv_text = data.get("text_content", "")
                    extracted_name = data.get("name", None)
            except Exception as ai_exc:
                logger.error(f"AI service extract-text error: {ai_exc}")
                raise HTTPException(status_code=500, detail=f"Failed to extract text from CV using AI service: {ai_exc}")
            
            if not cv_text or len(cv_text.strip()) < 50:
                raise HTTPException(
                    status_code=400,
                    detail="Could not extract sufficient text from CV. Please ensure the file is readable."
                )
            
            # Use provided name or extracted name as fallback
            final_name = candidate_name or extracted_name or "Unknown Candidate"
            
            # Create application with S3 file URL
            logger.info(f"Creating application with file_url: {file_url}")
            logger.info(f"Creating application with s3_key: {s3_key}")
            
            application = await self.database_service.create_application(
                job_id=job_id,
                cv_filename=file_url,  # Store the full URL (S3 or local)
                candidate_email=candidate_email,
                candidate_name=final_name
            )
            
            logger.info(f" Created application: {application.application_id}")
            
            # Start asynchronous CV evaluation workflow
            try:
                evaluation_result = await self.evaluation_service.evaluate_cv_for_job(
                    cv_text=cv_text,
                    job_posting=job_posting,
                    application_id=application.application_id,
                    candidate_email=candidate_email,
                    candidate_name=final_name
                )
                
                if evaluation_result:
                    logger.info(f" CV evaluation completed: {evaluation_result.decision}")
                else:
                    logger.warning(" CV evaluation failed but application was recorded")
                
            except Exception as eval_error:
                logger.error(f" CV evaluation error: {eval_error}")
                # Don't fail the application submission if evaluation fails
                await self.database_service.update_application_status(
                    application.application_id, "EVALUATION_FAILED"
                )
            
            # Prepare response
            response_data = {
                "application_id": application.application_id,
                "job_title": job_posting.title,
                "status": "submitted",
                "message": "Your application has been submitted successfully!",
                "next_steps": "Our system is evaluating your CV. You will receive an email with the results shortly."
            }
            
            if evaluation_result:
                response_data.update({
                    "evaluation_score": evaluation_result.score,
                    "evaluation_decision": evaluation_result.decision.value
                })
            
            return success_response(
                message="Application submitted successfully",
                data=response_data
            )
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f" Error processing application: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to process application: {str(e)}")
    
    async def delete_job_posting(self, job_id: str):
        """
        Delete a job posting and all its associated applications
        
        Args:
            job_id: Job posting ID to delete
            
        Returns:
            APIResponse: Deletion confirmation
        """
        try:
            logger.info(f" Deleting job posting: {job_id}")
            
            # Check if job posting exists
            job_posting = await self.database_service.get_job_posting_by_id(job_id)
            if not job_posting:
                raise HTTPException(
                    status_code=404,
                    detail="Job posting not found"
                )
            
            # Delete job posting and associated applications
            success = await self.database_service.delete_job_posting(job_id)
            
            if not success:
                raise HTTPException(
                    status_code=500,
                    detail="Failed to delete job posting"
                )
            
            logger.info(f" Successfully deleted job posting: {job_id}")
            
            return success_response(
                message="Job posting deleted successfully",
                data={
                    "job_id": job_id,
                    "title": job_posting.title,
                    "deleted_at": datetime.now().isoformat()
                }
            )
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f" Error deleting job posting: {e}")
            raise HTTPException(status_code=500, detail=f"Failed to delete job posting: {str(e)}") 

File: services/ats-system-checker/src/controllers/job.controller.ts
import { Request, Response } from "express";
import { DatabaseService } from "../services/database.service";
import { EvaluationService } from "../services/evaluation.service";
import { S3Service } from "../services/s3.service";
import { JobPostingResponse, JobPostingsListResponse, ApplicationResponse } from "../models/api.models";
import { EvaluationDecision } from "../models/evaluation.models";
import { Env } from "../config/env.config";
import { v4 as uuidv4 } from "uuid";
import crypto from "crypto";
import pdfParse from "pdf-parse";
import fs from "fs";
import path from "path";

export class JobController {
  private database_service: DatabaseService;
  private evaluation_service: EvaluationService;
  private s3_service: S3Service;

  constructor(
    database_service: DatabaseService,
    evaluation_service: EvaluationService,
    s3_service: S3Service
  ) {
    this.database_service = database_service;
    this.evaluation_service = evaluation_service;
    this.s3_service = s3_service;
  }

  async create_job_posting(req: Request, res: Response): Promise<Response> {
    try {
      const {
        title,
        description,
        required_skills,
        additional_details,
        hr_email,
        hr_name,
        owner_user_id,
        owner_username,
        evaluation_threshold = 70,
        quiz_required = true,
        quiz_pass_threshold = 7
      } = req.body;

      console.log(`Creating new job posting: ${title}`);

      if (!title?.trim() || !description?.trim() || !required_skills?.trim()) {
        return res.status(400).json({
          success: false,
          error: true,
          message: "Job title, description, and required skills are required"
        });
      }

      const skills_list = required_skills.split(",").map((s: string) => s.trim()).filter(Boolean);

      const job_posting = await this.database_service.create_job_posting(
        title,
        description,
        skills_list,
        additional_details,
        hr_email,
        hr_name,
        evaluation_threshold,
        quiz_required,
        quiz_pass_threshold,
        owner_user_id,
        owner_username
      );

      const base_url = Env.BACKEND_URL;
      const shareable_link = `${base_url}/apply/${job_posting.job_id}`;

      console.log(`Created job posting with ID: ${job_posting.job_id}`);

      const response: JobPostingResponse = {
        job_id: job_posting.job_id,
        title: job_posting.title,
        description: job_posting.description,
        required_skills: job_posting.required_skills,
        shareable_link,
        created_at: job_posting.created_at,
        is_active: job_posting.is_active,
        hr_email: job_posting.hr_email,
        hr_name: job_posting.hr_name,
        created_by: job_posting.owner_user_id,
        owner_username: job_posting.owner_username
      };

      return res.status(201).json(response);
    } catch (error: any) {
      console.error(`Error creating job posting: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to create job posting: ${error.message}`
      });
    }
  }

  async get_job_posting(req: Request, res: Response): Promise<Response> {
    try {
      const job_id = Array.isArray(req.params.job_id) ? req.params.job_id[0] : req.params.job_id;
      console.log(`Retrieving job posting: ${job_id}`);

      const job_posting = await this.database_service.get_job_posting_by_id(job_id);
      if (!job_posting) {
        return res.status(404).json({
          success: false,
          error: true,
          message: "Job posting not found"
        });
      }

      // Check if job is active - if not, only allow owner to view it
      const x_user_id = req.headers["x-user-id"] as string;
      if (!job_posting.is_active) {
        // If user is not the owner, return 404 (as if job doesn't exist)
        if (!x_user_id || job_posting.owner_user_id !== x_user_id) {
          return res.status(404).json({
            success: false,
            error: true,
            message: "Job posting not found"
          });
        }
      }

      const base_url = Env.BACKEND_URL;
      const shareable_link = `${base_url}/apply/${job_posting.job_id}`;

      const response: JobPostingResponse = {
        job_id: job_posting.job_id,
        title: job_posting.title,
        description: job_posting.description,
        required_skills: job_posting.required_skills,
        shareable_link,
        created_at: job_posting.created_at,
        is_active: job_posting.is_active,
        hr_email: job_posting.hr_email,
        hr_name: job_posting.hr_name,
        created_by: job_posting.owner_user_id,
        owner_username: job_posting.owner_username
      };

      return res.status(200).json(response);
    } catch (error: any) {
      console.error(`Error getting job posting: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to get job posting: ${error.message}`
      });
    }
  }

  async get_all_job_postings(req: Request, res: Response): Promise<Response> {
    try {
      console.log("Retrieving all job postings");
      
      const x_user_id = req.headers["x-user-id"] as string;
      const include_inactive = req.query.include_inactive === "true";
      
      const job_postings = await this.database_service.get_all_job_postings(x_user_id, include_inactive);
      const base_url = Env.BACKEND_URL;

      const job_responses: JobPostingResponse[] = job_postings.map(job => ({
        job_id: job.job_id,
        title: job.title,
        description: job.description,
        required_skills: job.required_skills,
        shareable_link: `${base_url}/apply/${job.job_id}`,
        created_at: job.created_at,
        is_active: job.is_active,
        hr_email: job.hr_email,
        hr_name: job.hr_name,
        created_by: job.owner_user_id,
        owner_username: job.owner_username
      }));

      const response: JobPostingsListResponse = {
        total_jobs: job_responses.length,
        jobs: job_responses
      };

      return res.status(200).json(response);
    } catch (error: any) {
      console.error(`Error getting job postings: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to get job postings: ${error.message}`
      });
    }
  }

  async update_job_posting(req: Request, res: Response): Promise<Response> {
    try {
      const job_id = Array.isArray(req.params.job_id) ? req.params.job_id[0] : req.params.job_id;
      const updates = req.body;

      console.log(`Updating job posting: ${job_id}`);

      const job_posting = await this.database_service.update_job_posting(job_id, updates);
      if (!job_posting) {
        return res.status(404).json({
          success: false,
          error: true,
          message: "Job posting not found"
        });
      }

      const base_url = Env.BACKEND_URL;
      const shareable_link = `${base_url}/apply/${job_posting.job_id}`;

      const response: JobPostingResponse = {
        job_id: job_posting.job_id,
        title: job_posting.title,
        description: job_posting.description,
        required_skills: job_posting.required_skills,
        shareable_link,
        created_at: job_posting.created_at,
        is_active: job_posting.is_active,
        hr_email: job_posting.hr_email,
        hr_name: job_posting.hr_name,
        created_by: job_posting.owner_user_id,
        owner_username: job_posting.owner_username
      };

      return res.status(200).json(response);
    } catch (error: any) {
      console.error(`Error updating job posting: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to update job posting: ${error.message}`
      });
    }
  }

  async delete_job_posting(req: Request, res: Response): Promise<Response> {
    try {
      const job_id = Array.isArray(req.params.job_id) ? req.params.job_id[0] : req.params.job_id;
      console.log(`Deleting job posting: ${job_id}`);

      const success = await this.database_service.delete_job_posting(job_id);
      if (!success) {
        return res.status(404).json({
          success: false,
          error: true,
          message: "Job posting not found"
        });
      }

      return res.status(200).json({
        success: true,
        message: "Job posting deleted successfully"
      });
    } catch (error: any) {
      console.error(`Error deleting job posting: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to delete job posting: ${error.message}`
      });
    }
  }

  async apply_for_job(req: Request, res: Response): Promise<Response> {
    try {
      const job_id = Array.isArray(req.params.job_id) ? req.params.job_id[0] : req.params.job_id;
      const { candidate_email, candidate_name } = req.body || {};
      const file = req.file;

      console.log(`Processing job application for job: ${job_id}`);

      if (!file) {
        return res.status(400).json({
          success: false,
          error: true,
          message: "CV file is required (cv_file)"
        });
      }

      if (!candidate_email || !String(candidate_email).trim()) {
        return res.status(400).json({
          success: false,
          error: true,
          message: "candidate_email is required so we can send you the evaluation result"
        });
      }

      // Get job posting
      const job_posting = await this.database_service.get_job_posting_by_id(job_id);
      if (!job_posting || !job_posting.is_active) {
        return res.status(404).json({
          success: false,
          error: true,
          message: "Job posting not found or inactive"
        });
      }

      // Upload file to S3
      const upload_result = await this.s3_service.upload_file(file);
      const s3_key = upload_result.s3_key;

      // Create application
      const application_id = uuidv4();
      const application = await this.database_service.create_application(
        application_id,
        job_id as string,
        s3_key,
        candidate_email,
        candidate_name
      );

      // Send "new application" email to job owner / HR so they receive the message when CV is uploaded
      const email_service = this.evaluation_service.get_email_service();
      const notify_email = job_posting.hr_email || Env.EMAIL_FROM;
      if (email_service && notify_email) {
        email_service.send_new_application_notification(
          notify_email,
          job_posting.title || "Job Position",
          candidate_name || "Candidate",
          candidate_email || "",
          application.application_id,
          Env.BACKEND_URL
        ).then(sent => {
          if (sent) console.log("New application notification sent");
          else console.warn("Failed to send new application notification");
        }).catch(err => console.error("Error sending new application email:", err.message));
      }

      // Process CV evaluation and send email notifications asynchronously (includes email to candidate with result)
      this._process_cv_evaluation_async(
        file,
        job_posting,
        application,
        candidate_email,
        candidate_name
      ).catch(error => {
        console.error("Error processing CV evaluation:", error);
      });

      const response: ApplicationResponse = {
        application_id: application.application_id,
        job_id: application.job_id,
        status: application.status,
        quiz_required: job_posting.quiz_required
      };

      return res.status(201).json(response);
    } catch (error: any) {
      console.error(`Error processing job application: ${error.message}`);
      return res.status(500).json({
        success: false,
        error: true,
        message: `Failed to process application: ${error.message}`
      });
    }
  }

  private async _process_cv_evaluation_async(
    file: Express.Multer.File,
    job_posting: any,
    application: any,
    candidate_email: string,
    candidate_name: string
  ): Promise<void> {
    try {
      const startTime = Date.now();
      
      // Extract text from CV file
      let cv_text = "";
      const file_extension = path.extname(file.originalname).toLowerCase();
      
      if (file_extension === ".pdf") {
        const pdf_buffer = fs.readFileSync(file.path);
        const pdf_data = await pdfParse(pdf_buffer);
        cv_text = pdf_data.text;
      } else if (file_extension === ".docx") {
        try {
          // Dynamic import to avoid loading issues
          const mammoth = await import("mammoth");
          const docx_buffer = fs.readFileSync(file.path);
          const result = await mammoth.default.extractRawText({ buffer: docx_buffer });
          cv_text = result.value;
        } catch (docxError: any) {
          console.error("Error extracting text from DOCX:", docxError.message);
          // If DOCX extraction fails, send email notification
          const email_service = this.evaluation_service.get_email_service();
          if (email_service && candidate_email) {
            try {
              await email_service.send_email_async(
                candidate_email,
                "CV Submission Received",
                `<p>Dear ${candidate_name || "Candidate"},</p>
                <p>Thank you for submitting your CV. We have received your application.</p>
                <p>However, we encountered an issue processing your DOCX file. Please resubmit your CV as a PDF file for automated evaluation.</p>
                <p>Best regards,<br/>Staffly Team</p>`,
                "CV_SUBMISSION"
              );
            } catch (emailError: any) {
              console.error(`Failed to send email:`, emailError.message);
            }
          }
          return;
        }
      } else {
        console.warn(`Unsupported file type: ${file_extension}`);
        return;
      }

      if (!cv_text || cv_text.trim().length === 0) {
        console.warn("No text extracted from CV file");
        const email_service = this.evaluation_service.get_email_service();
        if (email_service && candidate_email) {
          try {
            await email_service.send_email_async(
              candidate_email,
              "CV Received - Manual Review",
              `<p>Dear ${candidate_name || "Candidate"},</p>
              <p>Thank you for submitting your CV for <strong>${job_posting.title || "the position"}</strong>.</p>
              <p>We have received your application and saved your CV. We could not extract text from your file for automated evaluation, so your application will be reviewed manually.</p>
              <p>Best regards,<br/>Staffly Team</p>`,
              "CV_RECEIVED"
            );
          } catch (e: any) {
            console.error("Failed to send CV-received email:", e.message);
          }
        }
        return;
      }

      // Evaluate CV using AI service
      const evaluation_result = await this.evaluation_service.evaluate_cv(
        cv_text,
        job_posting.description || "",
        job_posting.required_skills || []
      );

      const processing_time = Date.now() - startTime;

      // Save evaluation to database
      const evaluation = await this.database_service.save_cv_evaluation(
        file.originalname,
        job_posting.description || "",
        evaluation_result.decision,
        evaluation_result.score,
        evaluation_result.evaluation_text,
        evaluation_result.text_length,
        candidate_email || evaluation_result.email,
        processing_time,
        job_posting.owner_user_id
      );

      // Update application with cv_score and decision so GET /applications/{id} and list views show correct data
      const isAccepted =
        evaluation_result.decision === EvaluationDecision.ACCEPT ||
        evaluation_result.decision === EvaluationDecision.ACCEPTED;
      await this.database_service.update_application(application.application_id, {
        cv_score: evaluation_result.score,
        decision: String(evaluation_result.decision),
        status: isAccepted ? "ACCEPTED" : "REJECTED"
      });

      // Generate quiz link if quiz is required and CV is accepted
      let quiz_link: string | undefined;
      if (job_posting.quiz_required && (evaluation_result.decision === EvaluationDecision.ACCEPT || evaluation_result.decision === EvaluationDecision.ACCEPTED)) {
        // Create quiz session
        const quiz = await this.evaluation_service.generate_quiz(job_posting.description || "");
        const quiz_session = await this.database_service.save_quiz_session(
          job_posting.description || "",
          quiz.questions || [],
          file.originalname,
          candidate_email || "",
          job_posting.quiz_time_limit || 300,
          job_posting.quiz_pass_threshold || 7,
          job_posting.owner_user_id
        );

        quiz_link = `${Env.FRONTEND_URL_QUIZ}/ats-checker/quiz/${quiz_session._id}`;
      }

      // Send email to the applicant with the AI evaluation result (score, decision, reasoning)
      const email_service = this.evaluation_service.get_email_service();
      const applicant_email = candidate_email || evaluation_result.email;
      if (email_service && applicant_email) {
        try {
          await email_service.send_cv_result_email(
            applicant_email,
            candidate_name || "Candidate",
            job_posting.title || "Job Position",
            String(evaluation_result.decision),
            evaluation_result.score,
            quiz_link,
            evaluation_result.evaluation_text
          );
          console.log("CV result email sent to applicant");
        } catch (emailError: any) {
          console.error("Failed to send email to applicant:", emailError.message);
        }
      } else {
        console.warn("Email service not available or candidate email missing");
      }
    } catch (error: any) {
      console.error("Error in CV evaluation process:", error);
      // Optionally send error notification email
    }
  }
}

